kanidmd/src/lib/valueset.rs

```diff
@@ -0,0 +1,210 @@
+use crate::prelude::*;
+// use hashbrown::HashSet;
+use std::borrow::Borrow;
+use std::collections::BTreeSet;
+use std::iter::FromIterator;
+
+pub struct ValueSet {
+    inner: BTreeSet<Value>,
+}
+
+impl Default for ValueSet {
+    fn default() -> Self {
+        ValueSet {
+            inner: BTreeSet::new(),
+        }
+    }
+}
+
+impl ValueSet {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    // insert
+    pub fn insert(&mut self, value: Value) -> bool {
+        // Return true if the element is new.
+        self.inner.insert(value)
+    }
+
+    // set values
+    pub fn set(&mut self, iter: impl Iterator<Item = Value>) {
+        self.inner.clear();
+        self.inner.extend(iter);
+    }
+
+    pub fn get<Q>(&self, value: &Q) -> Option<&Value>
+    where
+        Value: Borrow<Q> + Ord,
+        Q: Ord + ?Sized,
+    {
+        self.inner.get(value)
+    }
+
+    // delete a value
+    pub fn remove<Q>(&mut self, value: &Q) -> bool
+    where
+        Value: Borrow<Q> + Ord,
+        Q: Ord + ?Sized,
+    {
+        self.inner.remove(value)
+    }
+
+    pub fn contains<Q>(&self, value: &Q) -> bool
+    where
+        Value: Borrow<Q> + Ord,
+        Q: Ord + ?Sized,
+    {
+        self.inner.contains(value)
+    }
+
+    pub fn substring(&self, value: &PartialValue) -> bool {
+        self.inner.iter().any(|v| v.substring(value))
+    }
+
+    pub fn lessthan(&self, value: &PartialValue) -> bool {
+        self.inner.iter().any(|v| v.lessthan(value))
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    // We'll need to be able to do partialeq/intersect etc later.
+
+    pub fn iter(&self) -> Iter {
+        (&self).into_iter()
+    }
+
+    pub fn difference<'a>(&'a self, other: &'a ValueSet) -> Difference<'a> {
```

Comment:
Why we need the lifetime annotations here (and in other functions)?

Reply:
Because https://doc.rust-lang.org/std/collections/btree_set/struct.Difference.html needs a lifetime, and we need to then correctly set them all through the other elements so they are constrained correctly.